<h1>UK ROAD ACCIDENT DATA ANALYSIS</h1>
<h1>INCLUSIVE YEAR 2019 - 2022</h1>

<h3>Analyst : Ramon Bernal</h3>

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from scipy.stats import f_oneway

In [2]:
accidents = pd.read_csv('datasets\\accident_data.csv')

In [3]:
accidents

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [5]:
accidents.isnull()

Index  Accident_Severity  Accident Date  Latitude  Light_Conditions  \
0       False              False          False     False             False   
1       False              False          False     False             False   
2       False              False          False     False             False   
3       False              False          False     False             False   
4       False              False          False     False             False   
...       ...                ...            ...       ...               ...   
660674  False              False          False     False             False   
660675  False              False          False     False             False   
660676  False              False          False     False             False   
660677  False              False          False     False             False   
660678  False              False          False     False             False   

        District Area  Longitude  Number_of_Casualties  Number_of_Vehicles  \
0               False      False                 False               False   
1               False      False                 False               False   
2               False      False                 False               False   
3               False      False                 False               False   
4               False      False                 False               False   
...               ...        ...                   ...                 ...   
660674          False      False                 False               False   
660675          False      False                 False               False   
660676          False      False                 False               False   
660677          False      False                 False               False   
660678          False      False                 False               False   

        Road_Surface_Conditions  Road_Type  Urban_or_Rural_Area  \
0                         False      False                False   
1                         False      False                False   
2                         False       True                False   
3                         False      False                False   
4                         False       True                False   
...                         ...        ...                  ...   
660674                    False      False                False   
660675                    False      False                False   
660676                    False      False                False   
660677                    False      False                False   
660678                    False      False                False   

        Weather_Conditions  Vehicle_Type  
0                    False         False  
1                    False         False  
2                     True         False  
3                    False         False  
4                     True         False  
...                    ...           ...  
660674               False         False  
660675               False         False  
660676               False         False  
660677               False         False  
660678               False         False  

[660679 rows x 14 columns]

In [6]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [7]:
accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

options

- delete the row

- fill the empty cells with a value

In [8]:
accidents['Latitude'] = accidents['Latitude'].fillna(accidents['Latitude'].mean())

In [9]:
accidents['Longitude'] = accidents['Longitude'].fillna(accidents['Longitude'].mean())

In [10]:
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].fillna(accidents['Urban_or_Rural_Area'].mode()[0])

In [11]:
accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                       0
Light_Conditions               0
District Area                  0
Longitude                      0
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area            0
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [12]:
accidents['Accident Date'] = pd.to_datetime(accidents['Accident Date'], dayfirst = True, errors = 'coerce')

In [13]:
accidents.dtypes

Index                              object
Accident_Severity                  object
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                   object
District Area                      object
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions            object
Road_Type                          object
Urban_or_Rural_Area                object
Weather_Conditions                 object
Vehicle_Type                       object
dtype: object

In [14]:
accidents['Accident_Severity'] = accidents['Accident_Severity'].astype('category')

In [15]:
accidents['Light_Conditions'] = accidents['Light_Conditions'].astype('category')

In [16]:
accidents['District Area'] = accidents['District Area'].astype('category')

In [17]:
accidents['Road_Surface_Conditions'] = accidents['Road_Surface_Conditions'].astype('category')

In [18]:
accidents['Road_Type'] = accidents['Road_Type'].astype('category')

In [19]:
accidents['Urban_or_Rural_Area'] = accidents['Urban_or_Rural_Area'].astype('category')

In [20]:
accidents['Weather_Conditions'] = accidents['Weather_Conditions'].astype('category')

In [21]:
accidents['Vehicle_Type'] = accidents['Vehicle_Type'].astype('category')

In [22]:
accidents.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1>INSIGHTS</h1>

<h1>Insight No. 1</h1>
<h1>Which meteorological circumstances result in the most accidents?</h1>
<h1>Okay, no high winds cause the most accidents.</h1>

In [23]:
weatherCdt_accidents = accidents['Weather_Conditions'].value_counts()

In [24]:
weatherCdt_accidents

Weather_Conditions
Fine no high winds       520885
Raining no high winds     79696
Other                     17150
Raining + high winds       9615
Fine + high winds          8554
Snowing no high winds      6238
Fog or mist                3528
Snowing + high winds        885
Name: count, dtype: int64

<h1>Insight No. 1</h1>
<h2>Multi-vehicle crashes does not result in more casualties</h2>

<h1>Insight No. 2</h1>
<h1>Do the number of vehicles engaged change between urban and rural areas?</h1>
<h1>Yes, the number of automobiles engaged in accidents varies amongst them.</h1>
<h1>There is a link between the number of automobiles participating and urban versus rural locations.</h1>

In [25]:
severity_district = accidents[accidents['Accident_Severity'] == 'Fatal']['District Area'].value_counts()

In [26]:
severity_district

District Area
Birmingham                  105
Leeds                        93
Highland                     88
East Riding of Yorkshire     85
Bradford                     71
                           ... 
Havant                        2
Orkney Islands                2
Harlow                        1
Stevenage                     1
Clackmannanshire              1
Name: count, Length: 422, dtype: int64

<h1>Insight No. 3</h1>
<h1>According to reports, Birmingham has the most fatal accidents (105), followed by Leeds (93), and Highland (88).</h1>

<h1>Relationship between accident weight and number of vehicles</h1>

In [27]:
severity_vehicles = np.round(accidents.groupby('Accident_Severity')['Number_of_Casualties'].mean(), 0)

In [28]:
severity_vehicles

Accident_Severity
Fatal      2.0
Serious    1.0
Slight     1.0
Name: Number_of_Casualties, dtype: float64

<h1>Insight No. 4</h1>
<h1>Which district region has the greatest average casualty rate</h1>
<h1>Blaenau Gwent has the greatest average number of casualties.</h1>

<h1>Most often, vehicle types related to fatal accidents</h1>

In [29]:
districtarea_accidents = accidents['District Area'].value_counts()

In [30]:
districtarea_accidents

District Area
Birmingham            13491
Leeds                  8898
Manchester             6720
Bradford               6212
Sheffield              5710
                      ...  
Berwick-upon-Tweed      153
Teesdale                142
Shetland Islands        133
Orkney Islands          117
Clackmannanshire         91
Name: count, Length: 422, dtype: int64

<h1>Insight No. 5</h1>


<h1>Cars are most fatal accidents (6,577 registrations).</h1>

In [31]:
light_severity = accidents.groupby('Light_Conditions')['Accident_Severity'].value_counts()

In [32]:
light_severity

Light_Conditions             Accident_Severity
Darkness - lighting unknown  Slight                 5622
                             Serious                 794
                             Fatal                    68
Darkness - lights lit        Slight               108345
                             Serious               19130
                             Fatal                  1860
Darkness - lights unlit      Slight                 2138
                             Serious                 360
                             Fatal                    45
Darkness - no lighting       Slight                28651
                             Serious                7174
                             Fatal                  1612
Daylight                     Slight               419045
                             Serious               60759
                             Fatal                  5076
Name: count, dtype: int64

<h1>Insight No. 6</h1>
<h1>The majority of accidents occur during daylight hours, with 419,045 minor, 60,759 serious, and 5,076 fatalities reported.</h1>


<h1>Relationship between road type and accident severity.</h1>

In [33]:
rtype_severity = accidents.groupby('Road_Type')['Accident_Severity'].value_counts()

In [34]:
rtype_severity

Road_Type           Accident_Severity
Dual carriageway    Slight                85863
                    Serious               11746
                    Fatal                  1815
One way street      Slight                11809
                    Serious                1655
                    Fatal                    95
Roundabout          Slight                40185
                    Serious                3665
                    Fatal                   142
Single carriageway  Slight               415557
                    Serious               70059
                    Fatal                  6527
Slip road           Slight                 6381
                    Serious                 611
                    Fatal                    49
Name: count, dtype: int64

<h1>Insight No. 7</h1>
<h1>Which weather and light conditions lead to the highest number of casualties?</h1>
<h1>The highest number of casualties occur during daylight and fine weather without high winds.</h1>

<h1>Correlation of urban/rural classification and accident severity.</h1>

In [35]:
area_severity = accidents.groupby('Urban_or_Rural_Area')['Accident_Severity'].value_counts()

In [36]:
area_severity

Urban_or_Rural_Area  Accident_Severity
Rural                Slight               196077
                     Serious               37312
                     Fatal                  5601
Unallocated          Slight                   10
                     Serious                   1
                     Fatal                     0
Urban                Slight               367714
                     Serious               50904
                     Fatal                  3060
Name: count, dtype: int64

<h1>Insight No. 8</h1>
<h1>Urban areas have more accidents but fewer fatalities, with 367,714 minor, 50,904 serious, and 3,060 fatal accidents reported.
Rural areas have fewer accidents, but a higher fatality rate, with 196,077 minor, 37,312 serious, and 5,601 fatal accidents reported.
<1h2>

<h1>Weather conditions affect the number of casualties.</h1>

In [37]:
weather_casualties = np.round(accidents.groupby('Weather_Conditions')['Number_of_Casualties'].mean(), 1)

In [38]:
weather_casualties

Weather_Conditions
Fine + high winds        1.4
Fine no high winds       1.4
Fog or mist              1.5
Other                    1.4
Raining + high winds     1.4
Raining no high winds    1.4
Snowing + high winds     1.4
Snowing no high winds    1.3
Name: Number_of_Casualties, dtype: float64

<h1>Insight No. 9</h1>
<h1>Foggy or misty weather conditions have the highest casualty rate (1.5).</h1>

<h1>Most common road surface condition for Fatal accidents</h1>

<h1>The most common road surface condition that leads to fatal accidents</h1>
<h1>Fatal Accidents Occur Most on Dry Roads</h1>

In [39]:
severity_rtype = accidents[accidents['Accident_Severity'] == 'Fatal']['Road_Surface_Conditions'].mode()[0]

<h1>Insight No. 10</h1>
<h1>Relationship between road type and number of vehicles involved</h1>
<h1>Dual carriageways have the highest number of cars per accident, with two.</h1>

In [40]:
severity_rtype

'Dry'

In [41]:
rtype_vehicles = np.round(accidents.groupby('Road_Type')['Number_of_Vehicles'].mean(), 1)

In [42]:
rtype_vehicles

Road_Type
Dual carriageway      2.0
One way street        1.6
Roundabout            1.9
Single carriageway    1.8
Slip road             1.9
Name: Number_of_Vehicles, dtype: float64

<h1>Insight No. 11</h1>
<h1>Comparison of accident severity under various meteorological conditions.</h1>
<h1>Most accidents happen in "Fine no high winds" circumstances, with 441,739 minor, 72,046 serious, and 7,100 death incidents reported.</h1>

In [43]:
severity_weather = accidents.groupby('Weather_Conditions')['Accident_Severity'].value_counts()

In [44]:
severity_weather

Weather_Conditions     Accident_Severity
Fine + high winds      Slight                 7134
                       Serious                1245
                       Fatal                   175
Fine no high winds     Slight               441739
                       Serious               72046
                       Fatal                  7100
Fog or mist            Slight                 2963
                       Serious                 483
                       Fatal                    82
Other                  Slight                15184
                       Serious                1801
                       Fatal                   165
Raining + high winds   Slight                 8209
                       Serious                1261
                       Fatal                   145
Raining no high winds  Slight                69380
                       Serious                9468
                       Fatal                   848
Snowing + high winds   Slight            

<h1>Insight No. 12</h1>
<h1>District with the highest number of accidents during the day</h1>

In [45]:
district_daylight = accidents[accidents['Light_Conditions'] == 'Daylight']['District Area'].value_counts()

In [46]:
district_daylight

District Area
Birmingham            9667
Leeds                 6482
Manchester            4610
Bradford              4459
Sheffield             4211
                      ... 
Berwick-upon-Tweed     120
Teesdale               120
Shetland Islands        89
Orkney Islands          74
Clackmannanshire        68
Name: count, Length: 422, dtype: int64

<h1>Insight No. 13</h1>
<h1>The district with the highest number of accidents at night</h1>
<h1>Birmingham has the most nighttime accidents</h1>

In [47]:
district_nighttime = accidents[accidents['Light_Conditions'] != 'Daylight']['District Area'].value_counts()

In [48]:
district_nighttime

District Area
Birmingham            3824
Leeds                 2416
Manchester            2110
Westminster           1801
Bradford              1753
                      ... 
Orkney Islands          43
Clackmannshire          38
Berwick-upon-Tweed      33
Clackmannanshire        23
Teesdale                22
Name: count, Length: 422, dtype: int64

<h1>Insight No. 14</h1>
<h1>Clustering of accident-prone areas based on latitude and longitude</h1>
<h1>The highest accident count (45) occurs at latitude 52.949719, longitude -0.977611</h1>

In [49]:
long_lat = accidents[['Latitude', 'Longitude']].value_counts()

In [50]:
long_lat

Latitude   Longitude
52.949719  -0.977611    45
52.458798  -1.871043    35
53.083165  -0.816789    33
52.967634  -1.190861    31
52.938860  -1.216694    29
                        ..
51.608856  -0.698675     1
51.608845  -0.816519     1
51.608842  -3.334621     1
51.608833  -0.182514     1
60.757544  -0.862815     1
Name: count, Length: 570518, dtype: int64

<h1>Insight No. 15</h1>
<h1>Proportion of accidents occurring on weekends</h1>
<h1>The proportion of accidents happening on weekends is 11.84%</h1>

In [51]:
weekends = (accidents['Accident Date'].dt.dayofweek >= 5).mean() * 100

In [52]:
weekends

11.840697222100294

<h1>Insight No. 16</h1>
<h1>Proportion of accidents occurring on weekdays</h1>
<h1>The proportion of accidents happening on weekdays is 28.27%</h1>

In [53]:
weekdays = (accidents['Accident Date'].dt.dayofweek < 5).mean() * 100

In [54]:
weekdays

28.27061250622466

<h1>Insight No. 17</h1>
<h1>Road type with the highest number of fatal accidents</h1>
<h1>Single carriageways have the highest number of fatal accidents, with 6,527 fatalities</h1>

In [55]:
rtype_fatal = accidents[accidents['Accident_Severity'] == 'Fatal']['Road_Type'].value_counts()

In [56]:
rtype_fatal

Road_Type
Single carriageway    6527
Dual carriageway      1815
Roundabout             142
One way street          95
Slip road               49
Name: count, dtype: int64

<h1>Insight No. 18</h1>
<h1>Most common day of the week for accidents</h1>
<h1>Saturday has the highest number of accidents (43,164), followed by Wednesday (40,037) and Friday (39,822)</h1>

In [57]:
accident_day = accidents['Accident Date'].dt.day_name().value_counts()

In [58]:
accident_day

Accident Date
Saturday     43164
Wednesday    40037
Friday       39822
Thursday     39641
Tuesday      38714
Sunday       35065
Monday       28564
Name: count, dtype: int64

<h1>Insight No. 19</h1>
<h1>Average number of vehicles involved in accidents on dry vs. wet roads</h1>
<h1>The average number of vehicles involved in accidents is the same (1.8) for both dry and wet or damp roads</h1>

In [59]:
vehicles_roadcon = np.round(accidents.groupby('Road_Surface_Conditions')['Number_of_Vehicles'].mean(), 1)

In [60]:
vehicles_roadcon

Road_Surface_Conditions
Dry                     1.8
Flood over 3cm. deep    1.6
Frost or ice            1.6
Snow                    1.7
Wet or damp             1.8
Name: Number_of_Vehicles, dtype: float64

<h1>Insight No. 20</h1>
<h1>Maximum number of casualties in a single-vehicle accident.</h1>

In [61]:
casualties_single = accidents[accidents['Number_of_Vehicles'] == 1]['Number_of_Casualties'].max()

In [62]:
casualties_single

68

In [63]:
accidents.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1>Insight No. 21</h1>
<h1>A single-vehicle collision can result in up to 68 victims.</h1>
<h1>Which vehicle type has the most accidents?</h1>
<h1>Cars have the highest number of accidents.</h1>

In [64]:
vehicletype_accidents = accidents['Vehicle_Type'].value_counts()

In [65]:
vehicletype_accidents

Vehicle_Type
Car                                      497992
Van / Goods 3.5 tonnes mgw or under       34160
Bus or coach (17 or more pass seats)      25878
Motorcycle over 500cc                     25657
Goods 7.5 tonnes mgw and over             17307
Motorcycle 125cc and under                15269
Taxi/Private hire car                     13294
Motorcycle over 125cc and up to 500cc      7656
Motorcycle 50cc and under                  7603
Goods over 3.5t. and under 7.5t            6096
Other vehicle                              5637
Minibus (8 - 16 passenger seats)           1976
Agricultural vehicle                       1947
Pedal cycle                                 197
Data missing or out of range                  6
Ridden horse                                  4
Name: count, dtype: int64

<h1>Insight No. 22</h1>
<h1>In which area has the has the higher number of casualties?</h1>
<h1>Urban area has a higher number of casualties</h1>

In [66]:
area_casualties = accidents.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()

In [67]:
area_casualties

Urban_or_Rural_Area
Rural          353515
Unallocated        13
Urban          543040
Name: Number_of_Casualties, dtype: int64

<h1>Insight No. 23</h1>
<h1>What is the average number of vehicles in an accident?</h1>
<h1>The average number of vehicles in an accident is 1.8 </h1>

In [68]:
avg_vehicles = np.round(accidents['Number_of_Vehicles'].mean(), 1)

In [69]:
avg_vehicles

1.8

<h1>Insight No. 24</h1>
<h1>In which road area has the higher number of accidents?</h1>
<h1>Urban roads have a higher accidents than roads in rural areas.</h1>

In [70]:
accidents_per_area_road = accidents.groupby(['Urban_or_Rural_Area', 'Road_Type']).size().unstack().T

In [71]:
accidents_per_area_road

Urban_or_Rural_Area   Rural  Unallocated   Urban
Road_Type                                       
Dual carriageway      48715            1   50708
One way street         1193            0   12366
Roundabout            15545            1   28446
Single carriageway   168010            9  324124
Slip road              4294            0    2747

<h1>Insight No.25</h1>
<h1>Which day of the week has the highest number of accidents?</h1>

In [72]:
day_of_week_accidents = accidents.groupby("Index").size()

In [73]:
day_of_week_accidents.idxmax()

'2.01E+12'